### PR TITLE
Use PacketList directly when sending

### PR DESCRIPTION
**Pull request by cocosolos**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Couple edge case fixes if a pending packet ends up being null. Mainly though this makes it so we use the PacketList directly when preparing packets to send instead of making a copy and using that. We currently make a copy of the packet list, process those copies, then delete the real packets. This apparently causes some sync issues (#6829) that should be fixed now.

I'm not sure why exactly this issue seems to be consistent, so this may require some further review, but this seems to fix it.

## Steps to test these changes

Log in with 2 characters, go to the same zone, both characters can see each other.

### CODE REVIEW

File: src/map/entities/charentity.cpp [hunk 1]
```diff
@@ -439,40 +439,34 @@ void CCharEntity::pushPacket(std::unique_ptr<CBasicPacket>&& packet)
 void CCharEntity::updateCharPacket(CCharEntity* PChar, ENTITYUPDATE type, uint8 updatemask)
 {
     auto existing = PendingCharPackets.find(PChar->id);
-    if (existing == PendingCharPackets.end())
+    if (existing == PendingCharPackets.end() || !existing->second)
```
Comment (zach2good) [inline src/map/entities/charentity.cpp]: I find this logical block quite hard to reason about, can you break these up even if it ends up with nested conditions?

I think it would also help to rename existing to itr, and break out conditions like:

`const bool hasPendingPacket = itr != PendingCharPackets.end();`

and to not use the itr pair directly, but to use:

`auto& pendingPacket = existing->second;`
`pendingPacket ->updateWith...`

File: src/map/entities/charentity.h [hunk 2]
```diff
@@ -542,8 +541,7 @@ class CCharEntity : public CBattleEntity
 
     CItemEquipment* getEquip(SLOTTYPE slot);
 
-    // TODO: Don't use raw ptrs for this, but don't duplicate whole packets with unique_ptr either.
-    CBasicPacket* PendingPositionPacket = nullptr;
+    std::deque<std::unique_ptr<CBasicPacket>> PacketList; // The list of packets to be sent to the character during the next network cycle
```
Comment (zach2good) [inline src/map/entities/charentity.h]: instead of making this public, can you keep it private and access it with: `auto getPacketList() -> std::deque<std::unique_ptr<CBasicPacket>>&`

File: src/map/map.cpp [hunk 3]
```diff
@@ -1049,14 +1049,12 @@ int32 send_parse(int8* buff, size_t* buffsize, sockaddr_in* from, map_session_da
...
         do
         {
-            *buffsize       = FFXI_HEADER_SIZE;
-            auto packetList = PChar->getPacketListCopy();
-            packets         = 0;
+            *buffsize = FFXI_HEADER_SIZE;
+            packets   = 0;
 
-            while (!packetList.empty() && *buffsize + packetList.front()->getSize() < MAX_BUFFER_SIZE && static_cast<size_t>(packets) < PacketCount)
+            while (!PChar->PacketList.empty() && *buffsize + PChar->PacketList.front()->getSize() < MAX_BUFFER_SIZE && static_cast<size_t>(packets) < PacketCount)
```
Comment (zach2good) [inline src/map/map.cpp]: `isPacketListEmpty`

**Review by zach2good — APPROVED** [review]
Logically seems fine, I'd just like it to be made as clear and obvious about what's going on at every step as possible

File: src/map/entities/charentity.cpp (hunk 1 above)
Comment (cocosolos) [reply src/map/entities/charentity.cpp]: Updated the logic here, hopefully more clear.

File: src/map/entities/charentity.h (hunk 2 above)
Comment (cocosolos) [reply src/map/entities/charentity.h]: Good call, made it const too.

**Comment by BoundingS** [comment]
This seems to cause a very game breaking bug if items get out of sync between mog house and player inventory, gil etc.  Would you take a look at the issue I posted and see what you think?

**Comment by zach2good** [comment]
Please stop commenting on closed PRs, it's very hard for us to track. Open new issues and link prs you want to talk about
